kafkamocks/mock_producer.go

```diff
@@ -0,0 +1,113 @@
+package kafkamocks
+
+import (
+	"errors"
+
+	"github.com/Shopify/sarama"
+)
+
+type TestReporter interface {
+	Errorf(string, ...interface{})
+}
+
+type KafkaProducer interface {
+	AsyncClose()
+	Close() error
+	Input() chan<- *sarama.ProducerMessage
+	Successes() <-chan *sarama.ProducerMessage
+	Errors() <-chan *sarama.ProducerError
+}
+
+var (
+	errProduceSuccess    error = nil
+	errOutOfExpectations       = errors.New("No more expectations set on MockProducer")
+)
+
+type MockProducerExpectation struct {
+	Result error
+}
+
+type MockProducer struct {
+	expectations []*MockProducerExpectation
+	closed       chan struct{}
+	input        chan *sarama.ProducerMessage
+	successes    chan *sarama.ProducerMessage
+	errors       chan *sarama.ProducerError
+}
+
+func NewMockProducer(t TestReporter, config *sarama.Config) *MockProducer {
+	if config == nil {
+		config = sarama.NewConfig()
+	}
```

Comment:
We need to import sarama anyway for `sarama.ProducerMessage` and `sarama.ProducerError`, because we cannot efficiently capture these with interfaces. Also, we will never use the MockProducer (or a MockConsumer) inside Sarama itself; these are purely provided to make the lives of sarama users easier when it comes to testing.

Maybe the best solution is to keep mockBroker as a private type inside sarama, and if theres a need for it, add a public MockBroker type to this package as a copy?

** This reminds me: should we rename `Broker` in sarama to `BrokerConnection` or `Connection`?